spec/webauthn/public_key_credential_spec.rb

```diff
@@ -93,21 +93,41 @@
       end
     end
 
-    context "when clientExtensionResults is received" do
-      let(:public_key_credential) do
-        WebAuthn::PublicKeyCredentialWithAttestation.new(
-          type: type,
-          id: id,
-          raw_id: raw_id,
-          client_extension_outputs: { "appid" => "true" },
-          response: attestation_response
-        )
+    context "when clientExtensionResults" do
+      context "are not received" do
+        let(:public_key_credential) do
+          WebAuthn::PublicKeyCredentialWithAttestation.new(
+            type: type,
+            id: id,
+            raw_id: raw_id,
+            client_extension_outputs: nil,
+            response: attestation_response
+          )
+        end
+
+        it "works" do
+          expect(public_key_credential.verify(challenge)).to be_truthy
+
+          expect(public_key_credential.client_extension_outputs).to be_nil
```

Comment:
@grzuy @bdewater – 
I still have doubts in terms of what should we return in `client_extensions_outputs` in each case. Right now we are just returning the same object that is being passed to us. So, it will be something like:

clientExtensionResults | client_extension_outputs
------------ | -------------
`nil` | `nil`
`{}` | `{}`
`{ "appid" => "true" }` | `{ "appid" => "true" }`
`{ appid: true }` | `{ appid: true }`
`<ActionController::Parameters {"appid"=>true} permitted: false>` | `<ActionController::Parameters {"appid"=>true} permitted: false>`

I was thinking that we should consider doing some kind of processing in order to get a more consistent output, something like:

clientExtensionResults | client_extension_outputs
------------ | -------------
`nil` | `{}`
`{}` | `{}`
`{ "appid" => "true" }` | `{ appid: true }`
`{ appid: true }` | `{ appid: true }`
`<ActionController::Parameters {"appid"=>true} permitted: false>` | `{ appid: true }`

What do you guys think about this?

Reply:
Good question...

I don't think we should try making the gem be "too smart" about the semantics of the extensions by doing any data normalization right now.

I am not saying we would never add any normalization, but I think we should release like this, and consider some normalization in the future as extensions mature in the ecosystem.

I want to avoid getting it wrong and needing to re-implement normalization later, meaning a breaking change.

Thanks for raising the concern :-)